services/spar/test/Test/Spar/Sem/IdPSpec.hs

```diff
@@ -9,64 +10,237 @@ import Imports
+  Property
+prop_deleteGet =
+  prepropLaw @'[E.IdP] $ do
+    s <- arbitrary
+    pure
+      ( do
+          E.deleteConfig s
+          E.getConfig $ s ^. SAML.idpId,
+        do
+          E.deleteConfig s
+          pure Nothing
+      )
+
+prop_deleteDelete ::
+  PropConstraints r f =>
+  (forall a. Sem r a -> IO (f a)) ->
+  Property
+prop_deleteDelete =
+  prepropLaw @'[E.IdP] $ do
+    s <- arbitrary
+    pure
+      ( do
+          E.deleteConfig s
+          E.deleteConfig s,
+        do
+          E.deleteConfig s
+      )
+
+prop_storeGetByIssuer ::
+  PropConstraints r f =>
+  (forall a. Sem r a -> IO (f a)) ->
+  Property
+prop_storeGetByIssuer =
+  prepropLaw @'[E.IdP] $
+    do
+      s <- arbitrary
+      pure
+        ( do
+            E.storeConfig s
+            E.getIdByIssuerWithoutTeam $ s ^. idpMetadata . edIssuer,
+          do
+            E.storeConfig s
+            pure $ E.GetIdPFound $ s ^. idpId
+        )
+
+prop_setClear ::
+  PropConstraints r f =>
+  (forall a. Sem r a -> IO (f a)) ->
+  Property
+prop_setClear =
+  prepropLaw @'[E.IdP] $
+    do
+      replaced <- arbitrary
+      replacing <- arbitrary
+      pure
+        ( do
+            E.setReplacedBy replaced replacing
+            E.clearReplacedBy replaced,
+          do
+            E.clearReplacedBy replaced
+        )
+
+prop_getGet ::
+  PropConstraints r f =>
+  (forall a. Sem r a -> IO (f a)) ->
+  Property
+prop_getGet =
+  prepropLaw @'[E.IdP] $
+    do
+      idpid <- arbitrary
+      pure
+        ( do
+            liftA2 (,) (E.getConfig idpid) (E.getConfig idpid),
+          do
+            cfg <- E.getConfig idpid
+            pure (cfg, cfg)
+        )
+
+prop_getStore ::
+  PropConstraints r f =>
+  (forall a. Sem r a -> IO (f a)) ->
+  Property
+prop_getStore =
+  prepropLaw @'[E.IdP] $
+    do
+      idpid <- arbitrary
+      pure
+        ( do
+            E.getConfig idpid >>= maybe (pure ()) E.storeConfig,
+          do
+            pure ()
+        )
+
+prop_setSet ::
+  PropConstraints r f =>
+  (forall a. Sem r a -> IO (f a)) ->
+  Property
+prop_setSet =
+  prepropLaw @'[E.IdP] $
+    do
+      replaced <- arbitrary
+      replacing <- arbitrary
+      replacing' <- arbitrary
+      pure
+        ( do
+            E.setReplacedBy replaced replacing
+            E.setReplacedBy replaced replacing',
+          do
+            E.setReplacedBy replaced replacing'
+        )
+
+prop_setGet ::
+  PropConstraints r f =>
+  (forall a. Sem r a -> IO (f a)) ->
+  Property
+prop_setGet =
+  prepropLaw @'[E.IdP] $
+    do
+      replaced_id <- arbitrary
+      let replaced = E.Replaced replaced_id
+      replacing_id <- arbitrary
+      let replacing = E.Replacing replacing_id
+      pure
+        ( do
+            E.setReplacedBy replaced replacing
+            getReplacedBy replaced_id,
+          do
+            E.setReplacedBy replaced replacing
+            (Just replacing_id <$) <$> E.getConfig replaced_id
```

Comment:
...  and I'm not sure this is making too much sense.  What are you trying to test here?

Reply:
This shows that `getReplacedBy` actually gives you back what you replaced it by, right?

Reply:
Yes, but why the `getConfig`?  Why just `pure $ Just replacing_id`?

Reply:
Ah, yeah. Because this is a `Maybe (Maybe IdPId)` where the outer maybe depends on whether the thing exists in the first place.